### ISSUE TITLE
Libraries sometimes depend on tsickle decorator downleveling

### ISSUE THREAD

**Issue by devversion** [angular/angular#30586]
# 🐞 bug report

### Affected Package

 The issue is caused by package @angular/....

`@angular/compiler-cli` / `@angular/bazel`

### Is this a regression?

No

### Description

Angular applications which are built with NGC but aren't setting up `tsickle` for decorator downleveling (e.g. don't set the `annotationsAs: "static fields"`), can't be rendered on the server if DOM-specific types are referenced on class members which are decorated.

e.g.

```ts
@Directive(...)
export class MyDirective {
  @Input() triggerElement: HTMLElement;
}
```

This results in the following ES5 JavaScript output if `tsickle` is **not enabled**

```js
    __decorate([
        core_1.Input(),
        __metadata("design:type", HTMLElement)
    ], MyDirective.prototype, "triggerElement");
```
Notice how the `__decorate` call tries to access `HTMLELement` from a global context. This causes an exception when the following file is executed in NodeJS since DOM APIs are not available. We mock most of the DOM APIs with `domino` but providing these mocks globally so that such failures can be avoided is not common right now.

Building the project **with tsickle** fixes the problem automatically as decorators are downleveled into a static field while `tsickle` also discards the type reference to `HTMLElement` (as it is not relevant for Angular).

```js
MyDirective.propDecorators = {
  triggerElement: [{ type: core_1.Input }]
};
```

TL;DR: 

1) Everything currently works as expected within the CLI since it provides optimizations that removes the decorators anyway when using AOT mode or build-optimizer.
2) It becomes an issue if a project is not built with the CLI / and does not use build-optimizer. Previously tsickle was part of the NGC workflow **by default**.. so it now causes unexpected failures. Needs documentation? See: https://github.com/angular/angular/commit/29761ea5f8e4bae68442236f4a0abc1666487a35

**How is Angular Bazel affected?**

With Bazel it is not possible to opt-in for tsickle's decorator downleveling. Tsickle can be enabled by passing the `bazelOptions: {tsickle: true}`, but that comes with more tsickle transformations which aren't desired. e.g. type imports are rewritten to `goog.requireType`.

## 🔬 Minimal Reproduction

https://github.com/devversion/ngc-without-tsickle

## 🔥 Exception or Error
<pre><code>__metadata("design:type", HTMLElement)
                                  ^
ReferenceError: HTMLElement is not defined
</code></pre>

## 🌍  Your Environment

**Angular Version:**
<pre><code>Latest: 8.0.0-rc.4</code></pre>

**Anything else relevant?**

**Comment by Splaktar** [comment]
Is this work relevant here? https://github.com/angular/angular-cli/pull/14473

**Comment by alan-agius4** [comment]
Just an addition to this, I remember that @alexeagle & @vikerman had a discussion with the Google TypeScript team and it was brought up that outside of Google unless using Closure there shouldn't be the need for tsickle. However seeing the latest issues, the above and https://github.com/angular/angular-cli/pull/14473 maybe it is. 

Tsickle will be re-introduced when generating libraries with CLI, however of course users might still opt-out if they want to use `resolveJsonModule` for example see https://github.com/angular/tsickle/pull/922. the main reason for this not to go in was not to incease the scope of tsickle.

**Comment by devversion** [comment]
This issue is mostly about projects *not* using the CLI but depending on libraries which no longer aren't transformed by `tsickle`. Tsickle previously was part of the default workflow with `ngc` and now it no longer is enabled by default.

The components repository still opts-in for `tsickle` decorator downleveling but we aren't able to do this when switching to Bazel (as mentioned in the issue description).

To be clear: Yes. In my opinion `tsickle` always did more than just making output closure-compatible. It had the nice side-effect of decorator downleveling for Angular / and making it SSR compatible.

**Comment by alexeagle** [comment]
History: when we first needed decorator downleveling around Angular 2 timeframe, TS didn't have CustomTransformers. tsickle was the closest thing available so we added the functionality there.

Now, it doesn't make sense. tsickle should be "ts to closure" (that's where the name derives: ts2cl) and the decorator downleveling should be a build step. I wish it was baked right into tsc/ngc using CustomTransformer but I guess putting it in the webpack layer via angular/angular-cli#14473 is our short-term answer to keep forwardRef working in target=es2015

**Comment by devversion** [comment]
Totally agreed that `tsickle` should be `TypeScript to Closure` only. Also agreed that ideally the decorator downleveling would be part of NGC.. Similarly to how `ngtsc` already handles the decorator downleveling. 

I'm not sure if it's worth doing anything to make it work for non-CLI applications in the meanwhile since we still have the `opt-in` for tsickle right now (which could stay until ngtsc is the default AFAIK). Though what we miss right now is a temporary solution to make this work with Bazel. There is no way to only opt into the decorator downleveling from `tsickle`.

**Comment by clydin** [comment]
The `design:type` metadata is a result of the TS `emitDecoratorMetadata` option, which is only needed in JIT to support dependency injection and is separate from decorator downleveling/removal.  If JIT mode is not required then the TS option can be removed. ([relevant TS code](https://github.com/microsoft/TypeScript/blob/master/src/compiler/transformers/ts.ts#L1668))

Regardless, the custom transformer introduced in `@ngtools/webpack` for 8.0 should ideally be moved into ngc proper (with additional refactoring and cleanup -- the code was intended to minimize risk by leveraging the existing Tsickle code as much as possible).

**Comment by clydin** [comment]
Also of interest is that TypeScript has an alternative format for the metadata that uses arrow functions.  However, it is disabled via a hardcoded constant. https://github.com/microsoft/TypeScript/blob/master/src/compiler/transformers/ts.ts#L1659

**Comment by trotyl** [comment]
> The `design:type` metadata is a result of the TS `emitDecoratorMetadata` option, which is only needed in JIT to support dependency injection and is separate from decorator downleveling/removal. If JIT mode is not required then the TS option can be removed.

@clydin Even some transformer rewrites the parameter type to a value production, TypeScript will still remove its import, resulting in broken code, which happens in https://github.com/angular/tsickle/issues/635.

**Comment by clydin** [comment]
The root cause of this issue is the use of the TypeScript `emitDecoratorMetadata` option.  Tsickle is only relevant here because it removes the decorators.  This effectively disables the option as there is no "decorator metadata" to emit without the decorators.  Further, no transform of the parameter types is necessary outside of providing JIT mode support.  As such, if JIT mode support is not required then this problem is effectively void.

**Comment by devversion** [comment]
I looked into this again now that Ivy is the default. The issue still remains as initially described for View Engine. With the new Ivy compiler though, this problem is solved as expected. Ngtsc removes Angular decorators from source files and adds appropriate `setClassMetadata` calls (if needed). These calls to `setClassMetadata` never contain superfluous references, so it will run as expected in SSR.

https://github.com/angular/angular/blob/252e3e948781cdd2efb7af46ec2988add341b3d3/packages/compiler-cli/src/ngtsc/transform/src/transform.ts#L88-L89

https://github.com/angular/angular/blob/252e3e948781cdd2efb7af46ec2988add341b3d3/packages/compiler-cli/src/ngtsc/transform/src/transform.ts#L156-L162

Developers that build libraries using plain `tsc` will still face those issues, but that was an issue before too. tsickle never ran for plain `tsc` unless explicitly wired up.

I think this issue can be closed. In the current state, where libraries still publish in View Engine, the CLI, `ng-packagr`, and `angular/components` have custom tooling that fix this issue (see issue description for more details). Once ngtsc is used for library publishing, the issue will be fixed too.
